### PR TITLE
Fix unstable and ugly formatting for comments in destructuring patterns

### LINKED ISSUE #15707: Object destructuring in parameters with type and comments results in non-stable format loop
**Prettier 3.1.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAdKcAeAHCAnGAAgBM4AzAQwFcAbIsGigZycIDEIJDh1DDIoTGHipgY+ABTBCvPoTKcANLL4B6VYQBGFPIQC8MkNryoQyqHMLqtFAF639h7fdPm+AXyTtOASSgBLGABKbnd0dzMQCGwYf2gmZFAdPAgAdwAFHQQElAoaVIoATwTFIzwKMABrOBgAZQoAWzgAGX8MZEoaJjhSzXKqmtrsCraAc2RhKh6Qbob-CZFprGw4PH8m2DyAFVWoHX84HM7u0qYxmjgARSoIeA68k5AAKyZMWvOrm7ukY+mARy+cHSKWwORAzAAtBg4KRiJFhBR-DQxgBhCANBoUZDgmg0SJnKCjC4AQRgwn8mio8HSq1a7R+D2mAAsYA0aAB1JmBQ7DMBwWrZQL+ABugUK2LALEiwqmflIsFqYDWMWJUGItRghQu9y602wKW67PK2Gx+sOq2FcEibW6BGBFFGmJ1j2GeFt2M1KyYSv8MUi+raMHZ-mIMCZyAAHAAGUp4OAA-xx+2OrEM3WlGAUTTB0PhpAAJlKVG6WyzR0ZpTgDU0MNhzQohKoDrgHDwmLJY2x1HEIHc7iAA)

```sh
--parser typescript
```

**Input:**

```ts
export default class Foo {
  constructor({ 
    foo,
    // bar = "bar",
    // bazz = "bazz",
  }: FooInit) {}
}
```

**Output:**

```ts
export default class Foo {
  constructor({
    foo,
  } // bar = "bar",
  // bazz = "bazz",
  : FooInit) {}
}
```

**Expected behavior:**

I would expect no changes whatsoever to the original input:

```ts
export default class Foo {
  constructor({ 
    foo,
    // bar = "bar",
    // bazz = "bazz",
  }: FooInit) {}
}
```

**Notes**

This behaviour is actually fairly buggy, not only is the output not desirable, the output *isn't stable*, feeding the output back into prettier changes it again:

```ts
export default class Foo {
  constructor({
    foo, // bar = "bar",
  } // bazz = "bazz",
  : FooInit) {}
}
```

which feeding back in becomes (and will from now on alternate between this and the above printing):

```ts
export default class Foo {
  constructor({
    foo, // bar = "bar",
    // bazz = "bazz",
  }: FooInit) {}
}
```

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Fixes #15707

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
**Prettier pr-15708**
[Playground link](https://deploy-preview-15708--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAEwA6VOYD0hmARgIYBOeBxZ5AXngL5KYBiEEmAvJnAA8ADpQDcIADQgIQmAEt0yUFUoQA7gAUqCNMhDkANmvIBPXVNKVyYANZwYAZXIBbOABk5UOMgBmhtHAWVrb2DkLWngDmyDCUAK6BIAHOcjHxiYJCcJRyrrCGACrZUFRycLpIfgYBUmhRBnAAinEQ8L7+iQBWaAIO9U0tbZUdUgCOg3AaqkIV+mgAtF5wACYrkiCx5HIGUQDCEM7O5HqGBut1UJENAIIwsXKkcfAa2R5e7dWJABYwzgYA6l85PA0OEwHAHDpgXIAG7Akx6MBocwgGEJACSUFWsAcYBysmuWIcMBMDQ+NRAIggAX+ViEehE5WyMO8Uk8AUoMCm5EiR3JiXClA5ehJWTQeLksnWIk8MH+cmWMC+yAAHAAGKSUODjORa7m847DT5SGDkUjyxXKpAAJikcQCBTNFSqFLgzlIK1Wyzc5EucR5cE4lCOdyiJyeEBAzGYQA)

```sh
--parser typescript
```

**Input:**

```tsx
const {
  // bar
  // baz
}: Foo = expr;
```

**Output:**

```tsx
TypeError: Cannot read properties of undefined (reading 'comments')
    at ts (https://deploy-preview-15708--prettier.netlify.app/lib/plugins/estree.js:14:352)
    at ee (https://deploy-preview-15708--prettier.netlify.app/lib/plugins/estree.js:14:590)
    at Cp (https://deploy-preview-15708--prettier.netlify.app/lib/plugins/estree.js:14:9889)
    at https://deploy-preview-15708--prettier.netlify.app/lib/plugins/estree.js:14:777
    at Array.some (<anonymous>)
    at $o (https://deploy-preview-15708--prettier.netlify.app/lib/plugins/estree.js:14:769)
    at Nn (https://deploy-preview-15708--prettier.netlify.app/lib/standalone.js:30:2799)
    at Ht (https://deploy-preview-15708--prettier.netlify.app/lib/standalone.js:30:8022)
    at je (https://deploy-preview-15708--prettier.netlify.app/lib/standalone.js:30:7194)
    at qn (https://deploy-preview-15708--prettier.netlify.app/lib/standalone.js:30:11017)
```

**Review by fisker — APPROVED** [review]
Not sure about this, but overall good.

**Prettier pr-15708**
[Playground link](https://deploy-preview-15708--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAEwA6VOYBmEEeAvngPSWaQC2dCMeAvJnAB4AOATgNx48kKBmx4CxUlApRqtCAyZ4kmAGIlMbTrz4gANCAhcYAS3TJQAQx48IAdwAK1hGmQhLAGzuWAnq4MARjyWYADWcDAAypaMADImUHDIhJ5ocIHBYRGRXCEJAObIMDwArukgaXQmRaXl2nA8JoywngAqDVDWJnCuSCkeaQZoBR5wAIolEPDJqeUAVmgckSPjk9N9swYAjmtwDrZcve5oALSJcAAml-ogxZYmHgUAwgp0lm6eHjfDUPmjAIIwYomAIleAOBrxRIzAblAAWMDoHgA6nCTPA0LkwHBIi50SYAG7onxuMBofwgAllACSUCusEiYEaxn+dMiMB8oxhgxAvAgaWRwS4bl4PQaBKSBgSaR4MH2lnyb255VyPBlbg5XB6TJMxhuvASMGRJguMDhyAAHAAGAw8OA7Ex2+WK94bWEGGCWALG03mpAAJgMJTSrS9vX6PLgdAClyuF1ill+JQVcHUPDeQIKHzBEBAZDIQA)

```sh
--parser typescript
```

**Input:**

```tsx
const {
  foo
}
// comment
= expr;

const {
  foo
}
// comment
: Foo = expr;
```

**Output:**

```tsx
const { foo } =
  // comment
  expr;

const {
  foo,
  // comment
}: Foo = expr;

```

File: src/language-js/comments/handle-comments.js [hunk 1]
```diff
@@ -976,6 +977,44 @@ function handleLastUnionElementInExpression({
...
+      addTrailingComment(precedingNode, comment);
+    } else {
+      // const {
+      //   // bar
+      //   // baz
+      // }: Foo = expr;
+      addDanglingComment(enclosingNode, comment);
+    }
+    return true;
+  }
```
[the remark this answers is not in the document]
Comment (fisker) [reply src/language-js/comments/handle-comments.js:1015]: I guess the result will be https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEwA6VOYD0hmARgIYBOeBxZ5AXngL5KYBiEEmAvJnAA8ADpQDcIADQgIQmAEt0yUFUoQA7gAUqCNMhDkANmvIBPXVNKVyYANZwYAZXIBbOABk5UOMgBmhtHAWVrb2DkLWngDmyDCUAK6BIAHOcjHxiYJCcJRyrrCGACrZUFRycLpIfgYBUmhRBnAAinEQ8L7+iQBWaAIO9U0tbZUdUgCOg3AaqkIV+mgAtF5wACYrkiCx5HIGUQDCEM7O5HqGBut1UJENAIIwsXKkcfAa2R5e7dWJABYwzgYA6l85PA0OEwHAHDpgXIAG7Akx6MBocwgGEJACSUFWsAcYBysmuWIcMBMDQ+NRAIggAX+ViEehE5WyMO8Uk8AUoMCm5EiR3JiXClA5ehJWTQeLksnWIk8MH+cmWMC+yAAHAAGKSUODjORa7m847DT5SGDkUjyxXKpAAJikcQCBTNFSqFLgzlIK1Wyzc5EucR5cE4lCOdyiJyeEBAzGYQA

[the remark this answers is not in the document]
Comment (sosukesuzuki) [reply src/language-js/comments/handle-comments.js:1015]: Got following diff from tests:

```diff
- const {
-   // bar
-   // baz
- }: Foo = expr;
+ const {} // bar
+ // baz
+ : Foo = expr;
 
- const [
-   // bar
-   // baz
- ]: Foo = expr;
+ const [] // bar
+ // baz
+ : Foo = expr;
```

**Pull request merged**